python/cudf/cudf/tests/dataset_synthesizer.py

```diff
@@ -0,0 +1,138 @@
+# - Presence of null values
+# - Type of data
+# - Sorted
+# - Size in # of rows
+# - Size in # of columns
+class Parameters:
+    def __init__(
+        self, num_rows=2048, num_cols=4, column_params=default_column_params
+    ):
+        self.num_rows = num_rows
+        self.num_cols = num_cols
+        self.column_params = column_params
+
+
+# Synthesizes a new dataset and stores it in a Parquet file:
+def synthesize(filepath, parameters):
+    data = {}
+    columns_to_sort = []
+    for i in range(parameters.num_cols):
+        column_parameters = parameters.column_params[i]
+        vals = np.array(
+            [
+                column_parameters.ty()
+                for _ in range(column_parameters.cardinality)
+            ]
+        )
+
+        # Generate data for current column
+        data[str(i)] = np.array(
+            [
+                None
+                if np.random.rand() < column_parameters.null_frequency
+                else np.random.choice(vals)
+                for _ in range(parameters.num_rows)
+            ]
+        )
+
+        # Check if marked for sorting
+        if column_parameters.is_sorted:
+            columns_to_sort.append(str(i))
+
+    # Create DataFrame and sort columns appropriately
+    df = pd.DataFrame(data)
+    df = df.sort_values(columns_to_sort)
+
+    # Store in Parquet file
+    df.to_parquet(filepath, row_group_size=64)
+
+
+default = Parameters()
+
+# Focusing on the company type column...
+very_sparse_column_params = list(default_column_params)
+very_sparse_column_params[2].null_frequency = 0.8
+very_sparse_column_params[2].cardinality = 16
+very_sparse = Parameters(
+    num_rows=2048 << 4, column_params=very_sparse_column_params
+)
+
+# Focusing on the date column...
+dates_sorted_column_params = list(default_column_params)
+dates_sorted_column_params[0].is_sorted = True
+dates_sorted_column_params[0].cardinality = 2048
+dates_sorted = Parameters(
+    num_rows=2048, column_params=dates_sorted_column_params
+)
+high_cardinality_column_params = list(default_column_params)
+high_cardinality_column_params[0].is_sorted = True
+high_cardinality_column_params[0].cardinality = 2048
+high_cardinality = Parameters(
+    num_rows=2048, column_params=high_cardinality_column_params
+)
+
+# Focusing on the city column...
+low_cardinality_column_params = list(default_column_params)
+low_cardinality_column_params[1].is_sorted = True
+low_cardinality_column_params[1].cardinality = 20
+low_cardinality = Parameters(
+    num_rows=2048, column_params=low_cardinality_column_params
+)
+
+# Simple
+# This generates data for 2048 people regarding their age, city, industry,
+# and name. There should be 4 row groups of 512 rows each. Sorting is by age.
+simple_column_params = list(default_column_params)
+simple_column_params[0].ty = g.person.age
+simple_column_params[0].is_sorted = True
+simple_column_params[1].is_sorted = False
+simple = Parameters(num_rows=2048, column_params=simple_column_params)
```

Comment:
It would be good to somehow parameterize these to allow for easier tweaking of the generated datasets

Reply:
I'm not sure I follow. Do you mean - make it easier to specify `Parameters`, maybe using a dictionary instead of a complex `Parameters` struct, so that it's easier to, for example, use in parameterized tests?

Reply:
I made several changes to dataset generation.
- Renamed to `dataset_generator` and `generate`
- Added comments/docs
- Added parameter to `generate` function for providing format-specific parameters (although, currently only Parquet is supported)
- Removed constants from `dataset_generator`
- Created and used a single simple `Parameters` configuration in `test_parquet` where it is used

Based on discussion with Vukasin, it might be worth revisiting dataset generation to scope out use for fuzz tests and improve performance maybe in a later PR.